File: drills/ts/solutions/nc57_spiral_matrix__boundaries.ts
export function spiralOrder(matrix: number[][]): number[] {
  if (matrix.length === 0) return [];

  // Four boundaries closing in. Each side is walked and then retired, and the
  // two guards below are the ones everybody forgets: on a single remaining row
  // or column the bottom and top edges are the same edge, so walking both would
  // emit it twice.
  let top = 0;
  let bottom = matrix.length - 1;
  let left = 0;
  let right = matrix[0].length - 1;
  const out: number[] = [];

  while (top <= bottom && left <= right) {
    for (let c = left; c <= right; c++) out.push(matrix[top][c]);
    for (let r = top + 1; r <= bottom; r++) out.push(matrix[r][right]);
    if (top < bottom) for (let c = right - 1; c >= left; c--) out.push(matrix[bottom][c]);
    if (left < right) for (let r = bottom - 1; r > top; r--) out.push(matrix[r][left]);
    top++;
    bottom--;
    left++;
    right--;
  }

  return out;
}

File: drills/elixir/solutions/nc57_spiral_matrix__boundaries.exs
defmodule Solution do
  def spiral_order([]), do: []

  # Four boundaries closing in. Each side is walked and then retired, and the
  # two guards below are the ones everybody forgets: on a single remaining row
  # or column the bottom and top edges are the same edge, so walking both would
  # emit it twice.
  def spiral_order(matrix) do
    rows = List.to_tuple(Enum.map(matrix, &List.to_tuple/1))
    walk(rows, 0, length(matrix) - 1, 0, length(hd(matrix)) - 1, [])
  end

  defp walk(_rows, top, bottom, left, right, acc) when top > bottom or left > right,
    do: Enum.reverse(acc)

  defp walk(rows, top, bottom, left, right, acc) do
    acc = Enum.reduce(left..right//1, acc, fn c, acc -> [at(rows, top, c) | acc] end)

    acc =
      if top + 1 <= bottom,
        do: Enum.reduce((top + 1)..bottom//1, acc, fn r, acc -> [at(rows, r, right) | acc] end),
        else: acc

    acc =
      if top < bottom and left <= right - 1,
        do:
          Enum.reduce((right - 1)..left//-1, acc, fn c, acc -> [at(rows, bottom, c) | acc] end),
        else: acc

    acc =
      if left < right and top + 1 <= bottom - 1,
        do:
          Enum.reduce((bottom - 1)..(top + 1)//-1, acc, fn r, acc -> [at(rows, r, left) | acc] end),
        else: acc

    walk(rows, top + 1, bottom - 1, left + 1, right - 1, acc)
  end

  defp at(rows, r, c), do: rows |> elem(r) |> elem(c)
end
